docs: add TOML configuration reference (#209)
CerebriumAI/documentation: toml-reference/toml-reference.mdx

---
title: TOML Reference

> description: Complete reference for all configuration options available in Cerebrium's `cerebrium.toml` files.

Reviewer: ```suggestion
description: Complete reference for all parameters available in Cerebrium's default`cerebrium.toml` configuration file.
```

---

The configuration is organized into the following main sections:

> - **[cerebrium.deployment]** Core settings like project name, Python version, and file inclusion rules

Reviewer: ```suggestion
- **[cerebrium.deployment]** Core settings like app name, Python version, and file inclusion rules
```

- **[cerebrium.runtime.custom]** Custom web server settings and app startup behavior
- **[cerebrium.hardware]** Compute resources including CPU, memory, and GPU specifications
- **[cerebrium.scaling]** Auto-scaling behavior and replica management
- **[cerebrium.dependencies]** Package management for Python (pip), system (apt), and Conda dependencies

## Deployment Configuration

The `[cerebrium.deployment]` section defines core deployment settings.

| Option                | Type     | Default                | Description                                 |
| --------------------- | -------- | ---------------------- | ------------------------------------------- |

> | name                  | string   | required               | Name of your deployment                     |

Reviewer: ```suggestion
| name                  | string   | required               | Desired app name                   |
```

We should probably standardise our language here and use language users would use.

| python_version        | string   | "3.12"                 | Python version to use (3.10, 3.11, 3.12)    |

> | disable_auth          | boolean  | false                  | Disable authentication for endpoints        |

Reviewer: ```suggestion
| disable_auth          | boolean  | false                  | Disable default token-based authentication on app endpoints        |
```

| pre_build_commands    | string[] | []                     | Commands to run before dependencies install |
| docker_base_image_url | string   | "debian:bookworm-slim" | Base Docker image                           |

<Info>
  Changes to python_version or docker_base_image_url trigger full rebuilds since
  they affect the base environment.
</Info>

## Runtime Configuration

The `[cerebrium.runtime.custom]` section configures custom web servers and runtime behavior.

| Option               | Type     | Default  | Description                                  |
| -------------------- | -------- | -------- | -------------------------------------------- |
| port                 | integer  | required | Port the application listens on              |
| entrypoint           | string[] | required | Command to start the application             |
| healthcheck_endpoint | string   | ""       | HTTP path for health checks (empty uses TCP) |

<Info>
  The port specified in entrypoint must match the port parameter. All endpoints
  will be available at `https://api.cortex.cerebrium.ai/v4/{project - id}/
  {app - name}/your/endpoint`
</Info>

> ## Hardware Configuration

Reviewer: May be worth adding those hard limits we added to config here.

The `[cerebrium.hardware]` section defines compute resources.

| Option    | Type    | Default     | Description                          |
| --------- | ------- | ----------- | ------------------------------------ |
| cpu       | float   | required    | Number of CPU cores                  |
| memory    | float   | required    | Memory allocation in GB              |
| compute   | string  | "CPU"       | Compute type (CPU, AMPERE_A10, etc.) |
| gpu_count | integer | 0           | Number of GPUs                       |
| provider  | string  | "aws"       | Cloud provider                       |
| region    | string  | "us-east-1" | Deployment region                    |

<Warning>
  Memory refers to RAM, not GPU VRAM. Ensure sufficient memory for your
  workload.
</Warning>

## Scaling Configuration

The `[cerebrium.scaling]` section controls auto-scaling behavior.

| Option                | Type    | Default                   | Description                            |
| --------------------- | ------- | ------------------------- | -------------------------------------- |

> | min_replicas          | integer | 0                         | Minimum running instances              |

Reviewer: ```suggestion
| min_replicas          | integer | 0                         | Minimum number of running instances              |
```

> | max_replicas          | integer | 2                         | Maximum running instances              |

Reviewer: ```suggestion
| max_replicas          | integer | 2                         | Maximum number of running instances              |
```

> | replica_concurrency   | integer | 10                        | Concurrent requests per replica        |

Reviewer: Very Jargony
```suggestion
| replica_concurrency   | integer | 10                        | Number of requests each replica can process at once        |
```

| response_grace_period | integer | 3600                      | Grace period in seconds                |
| cooldown              | integer | 1800                      | Cooldown period between scaling events |

> | scaling_metric        | string  | "concurrency_utilization" | Metric for scaling decisions           |

Reviewer: May be worth adding in other options here

| scaling_target        | integer | 100                       | Target value for scaling metric        |

> | scaling_buffer        | integer | optional                  | Buffer capacity for scaling            |

Reviewer: Someone who doesn't know what this feature is may not understand what the exact jargon is. 

```suggestion
| scaling_buffer        | integer | optional                  | Excess number of running replicas available for immediate processing           |
```

- **[cerebrium.scaling]** Auto-scaling behavior and replica management
- **[cerebrium.dependencies]** Package management for Python (pip), system (apt), and Conda dependencies

## Deployment Configuration

The `[cerebrium.deployment]` section defines core deployment settings.

| Option                | Type     | Default                | Description                                                 |
| --------------------- | -------- | ---------------------- | ----------------------------------------------------------- |
| name                  | string   | required               | Desired app name                                            |
| python_version        | string   | "3.12"                 | Python version to use (3.10, 3.11, 3.12)                    |
| disable_auth          | boolean  | false                  | Disable default token-based authentication on app endpoints |
| include               | string[] | ["*"]                  | Files/patterns to include in deployment                     |
| exclude               | string[] | [".*"]                 | Files/patterns to exclude from deployment                   |

> | shell_commands        | string[] | []                     | Commands to run during build                                |

Reviewer 2: ```suggestion
| shell_commands        | string[] | []                     | Commands to run at the end of the build                                |
```

It ensures extra capacity is maintained above what the scaling metric suggests.

For example, with `min_replicas=0` and `scaling_buffer=3`, the system will maintain 3 replicas as baseline capacity.

</Info>

## Dependencies

### Pip Dependencies

The `[cerebrium.dependencies.pip]` section lists Python package requirements.

```toml
[cerebrium.dependencies.pip]
torch = "==2.0.0"      # Exact version
numpy = "latest"       # Latest version
pandas = ">=1.5.0"     # Minimum version
```

### APT Dependencies

The `[cerebrium.dependencies.apt]` section specifies system packages.

```toml
[cerebrium.dependencies.apt]
ffmpeg = "latest"
libopenblas-base = "latest"
```

### Conda Dependencies

The `[cerebrium.dependencies.conda]` section manages Conda packages.

```toml
[cerebrium.dependencies.conda]
cuda = ">=11.7"
cudatoolkit = "11.7"
```

### Dependency Files

The `[cerebrium.dependencies.paths]` section allows using requirement files.

```toml
[cerebrium.dependencies.paths]
pip = "requirements.txt"
apt = "pkglist.txt"
conda = "conda_pkglist.txt"
```

<Warning>
  Changes to APT or Conda dependencies trigger full rebuilds. Consider batching
  updates.

> </Warning>

Reviewer 2: @devin remove this
